enhancements/machine-api/control-plane-set.md

```diff
@@ -0,0 +1,603 @@
+introduced as a new second level operator within OpenShift to perform the operations described in this proposal.
+
+The new CRD will define the specification for managing (creating, adopting, updating) Machines for the Control Plane.
+The operator will be responsible for ensuring the desired number of Control Plane Machines are present within the
+cluster as well as providing update mechanisms akin to those seen in a StatefulSets and Deployments to allow rollouts
+of updated configuration to the Control Plane Machines.
+
+### User Stories
+
+- As a cluster administrator of OpenShift, I would like to be able to safely and automatically vertically resize my
+  control plane as and when the demand on the control plane changes
+- As a cluster administrator of OpenShift, I would like to be able to automatically recover failed Control Plane
+  Machines
+- As a cluster administrator of OpenShift, I would like to be able to make changes to the configuration of the
+  underlying hardware of my control plane and have these changes safely applied using immutable hardware concepts
+- As a cluster administrator of OpenShift, I would like to be able to control rollout of changes to my Control Plane
+  Machines such that I can test changes with a single replica before applying the change to all replicas
+- As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to scale down my
+  control plane before adding new Control Plane Machines with newer configuration, notably, my environment does not
+  have capacity to add additional Machines during updates
+
+### API Extensions
+
+We will introduce a new `ControlPlaneSet` CRD to the `machine.openshift.io/v1beta1` API group. It will be based on the
+spec and status structures defined below.
+
+```go
+type ControlPlaneSetSpec struct {
+    // Replicas defines how many Control Plane Machines should be
+    // created by this ControlPlaneSet.
+    // This field is immutable and cannot be changed after cluster
+    // installation.
+    // The ControlPlaneSet only operates with 3 or 5 node control planes,
+    // 3 and 5 are the only valid values for this field.
+    // +kubebuilder:validation:Enum:=3;5
+    // +kubebuilder:default:=3
+    // +kubebuilder:validation:Required
+    Replicas *int32 `json:"replicas"`
+
+    // Strategy defines how the ControlPlaneSet will update
+    // Machines when it detects a change to the ProviderSpec.
+    // +optional
+    Strategy ControlPlaneSetStrategy `json:"strategy,omitempty"`
+
+    // FailureDomains is the list of failure domains or
+    // availability zones in which the ControlPlaneSet should
+    // balance the Control Plane Machines.
+    // This will be injected into the ProviderSpec in the
+    // appropriate location for the particular provider.
+    // +kubebuilder:validation:MinLength:=1
+    // +kubebuilder:validation:Required
+    FailureDomains []string `json:"failureDomains"`
```

Comment:
can this mean that users put regions here? Looks like it encourages that. Cross-region etcd is not supported.